lib/codegen/src/legalizer/mod.rs

```diff
@@ -170,6 +171,74 @@ fn expand_cond_trap(
 
 /// Jump tables.
 fn expand_br_table(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    cfg: &mut ControlFlowGraph,
+    isa: &TargetIsa,
+) {
+    if isa.flags().jump_tables_enabled() {
+        expand_br_table_jt(inst, func, cfg, isa);
+    } else {
+        expand_br_table_conds(inst, func, cfg, isa);
+    }
+}
+
+/// Expand br_table to jump table.
+fn expand_br_table_jt(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    cfg: &mut ControlFlowGraph,
+    isa: &TargetIsa,
+) {
+    use ir::condcodes::IntCC;
+
+    let (arg, table) = match func.dfg[inst] {
+        ir::InstructionData::BranchTable {
+            opcode: ir::Opcode::BrTable,
+            arg,
+            table,
+        } => (arg, table),
+        _ => panic!("Expected br_table: {}", func.dfg.display_inst(inst, None)),
+    };
+
+    let table_size = func.jump_tables[table].len();
+    let table_is_fully_dense = func.jump_tables[table].fully_dense();
+    let addr_ty = isa.pointer_type();
+    let entry_ty = I32;
+
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    // Bounds check
+    let oob = pos
+        .ins()
+        .icmp_imm(IntCC::UnsignedGreaterThanOrEqual, arg, table_size as i64);
+    let fallthrough_ebb = pos.func.dfg.make_ebb();
+    pos.ins().brnz(oob, fallthrough_ebb, &[]);
+
+    let base_addr = pos.ins().jump_table_base(addr_ty, table);
+    let entry = pos
+        .ins()
+        .jump_table_entry(addr_ty, arg, base_addr, entry_ty.bytes() as u8, table);
+
+    // If the table isn't fully dense, zero-check the entry.
+    if !table_is_fully_dense {
+        pos.ins().brz(entry, fallthrough_ebb, &[]);
+    }
+
+    let addr = pos.ins().iadd(base_addr, entry);
+    pos.ins().indirect_jump_table_br(addr, table);
+
+    let ebb = pos.current_ebb().unwrap();
+    pos.insert_ebb(fallthrough_ebb);
```

Comment:
It looks like the fallthrough_ebb doesn't have any instructions in it. And, since Cranelift's `br_table` falls through in the default case, in theory there ought to be code here for splitting the current block at that point.

One option which might simplify things here would be to change Cranelift's `br_table` to be more like wasm's `br_table` and have an explicit "default" destination ebb, rather than just falling through. That way, it could have `is_terminator` set, and then you wouldn't need to create any new basic blocks.

Reply:
I don't have a strong opinion on this one. I'd probably default to just splitting the ebb rather than changing the IR.

But I'm in here changing things now, so if we're gonna do that this seems like a good time. Do you have an opinion?

Reply:
I think I'd prefer to change the IR, assuming that doesn't turn out to be too much of a detour here.

We'll have to add an Ebb to `BranchInfo::Table` and teach a few things how to handle that. And update lib/wasm/src/code_translator.rs, though that looks straightforward because currently it always does a jump after doing the br_table, so we can just get rid of that jump and add the ebb as the default on the br_table. And in lib/frontend/src/frontend.rs, declare the extra control edge only if it isn't already declared. At a quick scan, that should be the interesting parts. If you want, I can help out with some of this too.

Reply:
No problem! I'll take a stab at it and shout if I need help.